File: rtc4j/src/main/java/com/rapidtable/sdk/rtc4j/resource/report/ReportDeleteObjectRequest.java
/*
 * Copyright Rapid Table, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License. A copy of the License is located at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

package com.rapidtable.sdk.rtc4j.resource.report;

import com.rapidtable.sdk.rtc4j.resource.IDeleteRequest;
import com.rapidtable.sdk.rtc4j.resource.PathConfig;

import java.io.IOException;
import java.util.Objects;

import static com.rapidtable.sdk.rtc4j.resource.PathConfig.OBJECT_PATH_PATTERN;

public class ReportDeleteObjectRequest implements IDeleteRequest {
    private final String path;

    ReportDeleteObjectRequest(final String path) {
        this.path = path;
    }

    public String getPath() {
        return path;
    }

    @Override
    public String getQuery() {
        return null;
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private String workspaceId = null;
        private String projectId = null;
        private String reportId = null;
        private String objectId = null;
        private String target = null;

        public ReportDeleteObjectRequest build() throws IOException {
            if (Objects.nonNull(target)) {
                final var matcher = OBJECT_PATH_PATTERN.matcher(target);
                if (!matcher.matches()) {
                    throw new IllegalArgumentException();
                }
                final var workspaceId = matcher.group("wid");
                final var projectId = matcher.group("pid");
                final var reportId = matcher.group("rid");
                final var objectId = matcher.group("oid");
                final var path = PathConfig.ROOT + PathConfig.WORKSPACE + String.format("/%s", workspaceId) +
                    PathConfig.PROJECTS + String.format("/%s", projectId) +
                    PathConfig.REPORTS + String.format("/%s", reportId) +
                    PathConfig.OBJECTS + String.format("/%s", objectId);

                return new ReportDeleteObjectRequest(path);
            } else {
                final var path = PathConfig.ROOT + PathConfig.WORKSPACE + String.format("/%s", workspaceId) +
                    PathConfig.PROJECTS + String.format("/%s", projectId) +
                    PathConfig.REPORTS + String.format("/%s", reportId) +
                    PathConfig.OBJECTS + String.format("/%s", objectId);

                return new ReportDeleteObjectRequest(path);
            }
        }

        public Builder workspaceId(final String workspaceId) {
            this.workspaceId = workspaceId;
            return this;
        }

        public Builder projectId(final String projectId) {
            this.projectId = projectId;
            return this;
        }

        public Builder reportId(final String reportId) {
            this.reportId = reportId;
            return this;
        }

        public Builder objectId(final String objectId) {
            this.objectId = objectId;
            return this;
        }

        public Builder target(final String target) {
            this.target = target;
            return this;
        }
    }
}

File: rtc4js/src/resource/report/report-delete-object-request.ts
/*
 * Copyright Rapid Table, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the 'License').
 * You may not use this file except in compliance with the License. A copy of the License is located at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import { IDeleteRequest } from '../delete-request.interface';
import { PathConfig } from '../path-config';

export class ReportDeleteObjectRequest implements IDeleteRequest {

    constructor(public path: string) {
        this.path = path;
    }

    public getPath(): string {
        return this.path;
    }

    public getQuery(): { [key: string]: string | number } {
        return {};
    }

    public static builder(): Builder {
        return new Builder();
    }

}

class Builder {
    private _workspaceId: string | null = null;
    private _projectId: string | null = null;
    private _reportId: string | null = null;
    private _objectId: string | null = null;
    private _target: string | null = null;

    public build(): ReportDeleteObjectRequest {
        if (this._target) {
            const matcher = PathConfig.REPORT_OBJECT_PATH_PATTERN.exec(this._target);
            if (!matcher) {
                throw new Error('IllegalArgumentException');
            }

            const workspaceId = matcher.groups?.wid;
            const projectId = matcher.groups?.pid;
            const reportId = matcher.groups?.rid;
            const objectId = matcher.groups?.oid;
            const path = PathConfig.ROOT + PathConfig.WORKSPACE + `/${workspaceId}` +
                PathConfig.PROJECTS + `/${projectId}` +
                PathConfig.REPORTS + `/${reportId}` +
                PathConfig.OBJECTS + `/${objectId}`;

            return new ReportDeleteObjectRequest(path);
        } else {
            const path = PathConfig.ROOT + PathConfig.WORKSPACE + `/${this._workspaceId}` +
                PathConfig.PROJECTS + `/${this._projectId}` +
                PathConfig.REPORTS + `/${this._reportId}` +
                PathConfig.OBJECTS + `/${this._objectId}`;

            return new ReportDeleteObjectRequest(path);
        }
    }

    public workspaceId(workspaceId: string): Builder {
        this._workspaceId = workspaceId;
        return this;
    }

    public projectId(projectId: string): Builder {
        this._projectId = projectId;
        return this;
    }

    public reportId(reportId: string): Builder {
        this._reportId = reportId;
        return this;
    }

    public objectId(objectId: string): Builder {
        this._objectId = objectId;
        return this;
    }

    public target(target: string): Builder {
        this._target = target;
        return this;
    }
}